// @kodadot1/metasquid/dist/shared/metasquid.26b5e105.cjs
'use strict';

const scule = require('scule');

function eventFrom(interaction, { blockNumber, caller, timestamp }, meta, currentOwner) {
  return {
    interaction,
    blockNumber: BigInt(blockNumber),
    caller,
    currentOwner: currentOwner ?? caller,
    timestamp,
    meta
  };
}
function toBaseCall(context) {
  const caller = "";
  const { blockNumber, timestamp } = toBaseBlock(context);
  return { caller, blockNumber, timestamp };
}
function ensure(value) {
  return value;
}
function metadataOf({ metadata }) {
  return metadata ?? "";
}
function toBaseBlock(context) {
  const blockFrom = () => {
    if ("block" in context) {
      return context.block;
    }
    if ("header" in context) {
      return context.header;
    }
    return context;
  };
  const block = blockFrom();
  const blockNumber = block.height.toString();
  const timestamp = new Date(block.timestamp);
  return { blockNumber, timestamp };
}
function toMap(array) {
  return new Map(array.map((item) => [item.id, item]));
}
function toEntityId(item) {
  return item.id;
}
function toUniqueSet(array) {
  return new Set(array.map(toEntityId));
}
function toEntity(entityConstructor, el) {
  const entity = new entityConstructor();
  for (const prop in el) {
    entity[scule.camelCase(prop)] = el[prop];
  }
  return entity;
}
function takeFirst(list) {
  return list.at(0);
}

exports.ensure = ensure;
exports.eventFrom = eventFrom;
exports.metadataOf = metadataOf;
exports.takeFirst = takeFirst;
exports.toBaseBlock = toBaseBlock;
exports.toBaseCall = toBaseCall;
exports.toEntity = toEntity;
exports.toEntityId = toEntityId;
exports.toMap = toMap;
exports.toUniqueSet = toUniqueSet;
